cityofaustin/atd-vz-data
# Replace Lambda function blueform_location_from_latlon with trigger

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/13057

This PR updates `location_id` as a stored generated column. I'm already thinking about deployment strategy because some of these DB operations take a really long time locally (probably around 1 hour total). We can brainstorm some ideas here in this PR. I also created https://github.com/cityofaustin/atd-data-tech/issues/13275 to handle the AWS and Github Actions cleanup after all these triggers are replaced.

Questions:
- ~There are a lot of materialized views affected by this change. Do we think it is worth ~inventorying~ _assessing_ them in case some of them are no longer used? Do we know if Xavier uses any of them? Could be a backlog issue but it could also save deployment time if `five_year_surface_polygons_with_crash_data` isn't used anymore.~ _Update: See below for discussion and decision to drop these materialized views since they are not used by the VZ team or us._
- Are we ok with SQS build failures until we clean everything up in the issue linked above?

## Testing
**URL to test:** 
Local

**Steps to test:**
To test this, start up your local VZDB instance by replicating the latest DB dump and then follow the steps below. ~Heads up that some of the steps will take a long time locally to test!~

### Test steps
1. On your local VZ instance, follow the deployment steps below except for the last step of turning off the ETLs.
2. Afterwards, you can see the insert trigger working by running the sql statements below on your local DB
```sql
-- insert a blueform crash with coordinates that will associate to location_id 8337406736
INSERT INTO atd_apd_blueform(latitude, longitude, date, case_id)
VALUES (30.16841600, -97.78449000, '2023-08-14', '00000001');

-- insert a blueform crash with coordinates that will not associate to a location
INSERT INTO atd_apd_blueform(latitude, longitude, date, case_id)
VALUES (30.386136, -97.67309, '2023-08-14', '00000002');

-- insert a blueform crash without coordinates that will not associate to a location
INSERT INTO atd_apd_blueform(latitude, longitude, date, case_id)
VALUES (NULL, NULL, '2023-08-14', '00000003');
```
3. Update an existing record with a `location_id` of `NULL` to have the latitude and longitude of a record with a location. After update, you should see the record associate with the same location as the record you took the lat/lon from.
4. Update an existing record with a `location_id` that is not null and update it with coordinates from an existing record that has latitude and longitude and no `location_id` associated.
5. Update an existing record with a `location_id` that is not null and update it with `NULL` for both latitude and longitude. You should see the record `location_id` update to `NULL`.

### Deployment steps
The DB changes were applied through the Hasura CLI with the migrations in the this diff.
~- [ ] Remove Hasura event trigger called `blueform_location_from_latlon`~
~- [ ] Run SQL in [update_noncr3_location.sql](https://github.com/cityofaustin/atd-vz-data/blob/md-13057-blueform-lat-lon-trigger/atd-vzd/triggers/update_noncr3_location.sql)~
~- [ ] Run SQL in [migration_noncr3_location_2023_08_02--1648.sql](https://github.com/cityofaustin/atd-vz-data/blob/md-13057-blueform-lat-lon-trigger/atd-vzd/migrations/migration_noncr3_location_2023_08_02--1648.sql)~
~- [ ] Switch off non-cr3 crash location housekeeping Airflow ETLs~

---
#### Ship list
